### ISSUE TITLE
Is there any magic?

### ISSUE THREAD

**Issue by kevinxxq** [ent/ent#1940]
Hello Team,
Thank you for the project and I am trying to work with it.

My question is for Field.Json(). In the docs, it is as below:
```
func field.JSON(name string, typ interface{}) *field.jsonBuilder
field.JSON on pkg.go.dev

JSON returns a new Field with type json that is serialized to the given object. For example:

field.JSON("dirs", []http.Dir{}).
    Optional()

field.JSON("info", &Info{}).
    Optional()
```

I have tried both []http.Dir{} and []*http.Dir{}, they are both working.

So there is a code block inside to know the element is a object or pointer, right?
I have tried to look into the source code but no finding for my knowledge limit.

Thank you.

**Comment by yonidavidson** [comment]
Hi @kevinxxq, not sure I fully understand your question,
can you please create a small example project with the 2 types and what is the question?

Thanks.

**Comment by a8m** [comment]
Thanks for helping with this @yonidavidson 🙏 

@kevinxxq  - a code example with your question can also work here.

**Comment by masseelch** [comment]
field.JSON uses `encoding/json`, which in turn uses `reflect` / `json.Marshaler` to serialize / deserialize from and to the database if you mean that. And yes, reflection always is "magic" ...

**Comment by kevinxxq** [comment]
@yonidavidson @a8m Thank you. I can define []customtype or []*customtype which means I can pass instances or pointers. My question is how the serialize /de serialize know what it is(instance or pointers). 

example code:
In schema section: 
I defined: 
field.JSON("dirs", []http.Dir{}).Optional()

after ent generated the codes, there will be a method: 
```
func (pc *ParentCreate) SetChildSpectrumData(psd []http.Dir) *ParentCreate {
	pc.mutation.SetDirs(psd)
	return pc
}
```

when I call it, I need pass []http.Dir to it.

On the other hand, 
I defined: 
field.JSON("dirs", []*http.Dir{}).Optional()

after ent generated the codes, there will be a method: 
```
func (pc *ParentCreate) SetChildSpectrumData(psd []*http.Dir) *ParentCreate {
	pc.mutation.SetDirs(psd)
	return pc
}
```

when I call it, I need pass []*http.Dir to it.

Just as @masseelch replied as above, it uses reflection internally.
Thank you All.
